### ISSUE TITLE
ReconnectingClientFactory should be parameterized

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3859]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#3859|
|Type|enhancement|
|Created|2009-05-28 14:20:02Z|
|Branch|https://github.com/twisted/twisted/tree/reconnect-clock-3859|

Right now it imports the reactor inside retry, which makes it impossible to test.

<details><summary>Searchable metadata</summary>

```
trac-id__3859 3859
type__enhancement enhancement
reporter__therve therve
priority__normal normal
milestone__ 
branch__branches_reconnect_clock_3859 branches/reconnect-clock-3859
branch_author__therve therve
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1243520402000000 1243520402000000
changetime__1244650481000000 1244650481000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_1"></a>|@therve commented|
|-|-|

(In [26980]) Branching to 'reconnect-clock-3859'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve removed owner|
|-|-|

Please review!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @therve|
|-|-|

  1. epytext for `ReconnectingClientFactory.clock` is malformed.
  1. Also, I think it'd be good to explicitly point out what happens to the clock if the factory is serialized.
  1. Also, a direct unit test for serialization/deserialization would be nice as well.

Please merge when those are done.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve removed owner|
|-|-|

OK, I did those things, but realized I can't use `__init__` as it will break compatibility. So I used a property: it's not a huge change but it deserves another look I think. Thanks!

**Comment by twisted-trac** [comment]
|<img alt="amaury's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|amaury commented|
|-|-|

This won't work when the Factory is a new-style class, which may happen with multiple inheritance.
For example, I added this near the top of test_factories.py, and tests fail:
```python
class ReconnectingClientFactory(ReconnectingClientFactory, object):
    pass
```
With new-style classes the `property()` has priority over the object's `__dict__`, it is necessary to add a setter. The following code seems to work with both styles of classes:
```python
    def _getClock(self):
        try:
            return self.__dict__['clock']
        except KeyError:
            from twisted.internet import reactor
            self.__dict__['clock'] = reactor
            return reactor
    def _setClock(self, clock):
        self.__dict__['clock'] = clock
    clock = property(_getClock, _setClock)
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve commented|
|-|-|

Thanks for the test, I ended up removing the property.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @therve|
|-|-|

The last assert in `test_serializeWithClock` is redundant, isn't it?

Good catch on the `__init__` thing.  I'd forgotten about that restriction completely. :(  This code is so gross, it makes me sad.  Since there doesn't seem to be any escaping from that, though, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve set status to `closed`|
|-|-|

(In [26998]) Merge reconnect-clock-3859

Author: therve
Reviewer: exarkun
Fixes [#3859](https://github.com/twisted/twisted/issues/3859)

Add a clock attribute to ReconnectingClientFactory, allowing to parametrized
callLater calls in tests.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
